core/src/WsTrust/MexParser.cs

```diff
@@ -0,0 +1,280 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+using Microsoft.Identity.Core.Http;
+
+namespace Microsoft.Identity.Core.WsTrust
+{
+    internal enum WsTrustVersion
+    {
+        WsTrust13,
+        WsTrust2005
+    }
+
+    internal class WsTrustAddress
+    {
+        public Uri Uri { get; set; }
+
+        public WsTrustVersion Version { get; set; }
+    }
+
+    internal class MexPolicy
+    {
+        public WsTrustVersion Version { get; set; }
+
+        public string Id { get; set; }
+
+        public UserAuthType AuthType { get; set; }
+
+        public Uri Url { get; set; }
+    }
+
+    internal static class MexParser
+    {
+        private const string WsTrustSoapTransport = "http://schemas.xmlsoap.org/soap/http";
+
+        public static async Task<WsTrustAddress> FetchWsTrustAddressFromMexAsync(string federationMetadataUrl, UserAuthType userAuthType, RequestContext requestContext)
+        // This method would return null if the wanted policy is not found in mexDocument. It could potentially throw XmlException or http-relevant exceptions too.
+        {
+            XDocument mexDocument = await FetchMexAsync(federationMetadataUrl, requestContext).ConfigureAwait(false);
+            return ExtractWsTrustAddressFromMex(mexDocument, userAuthType, requestContext);
+        }
+
+        internal static async Task<XDocument> FetchMexAsync(string federationMetadataUrl, RequestContext requestContext)
+        {
+            var httpResponse = await HttpRequest.SendGet( // It would potentially throw MsalServiceException
+                new UriBuilder(federationMetadataUrl).Uri, null, requestContext).ConfigureAwait(false);
+            if (httpResponse.StatusCode != System.Net.HttpStatusCode.OK) {
+                throw new Client.MsalException(
+                    MsalError.AccessingWsMetadataExchangeFailed,
+                    string.Format(CultureInfo.CurrentCulture,
+                        "Response status code does not indicate success: {0} ({1}).",
+                        (int)httpResponse.StatusCode, httpResponse.StatusCode));
+            }
+            return XDocument.Parse(httpResponse.Body, LoadOptions.None);
+        }
+
+        internal static WsTrustAddress ExtractWsTrustAddressFromMex(XDocument mexDocument, UserAuthType userAuthType, RequestContext requestContext)
```

Comment:
>ExtractWsTrustAddressFromMex [](start = 39, length = 28)

ExtractWsTrustAddressFromMexPolicy ?

Reply:
This might be subjective but I would tend to think that the policy concept does not apply to this signature, instead it happens only inside this method. So we do not mention policy here.

---
In reply to: [201819995](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1118#discussion_r201819995) [](ancestors = 201819995)